src/roswire/ros2/reader.py

```diff
@@ -0,0 +1,141 @@
+        files.copy_from_host(host_script,
+                             '/launch_extractor.py')
+
+        # Runs the script using app_instance.shell.
+        output = shlex.quote(os.path.basename(filename) + '.json')
+        cmd = f'python3 /launch_extractor.py --output' \
+              f' {output} {shlex.quote(filename)}'
+        logger.debug(f"Running the script in the container: {cmd}")
+        self.app_instance.shell.check_call(cmd)
+        logger.debug(f"Reading {output} on container")
+        config_json = files.read(output)
+        config_nodes = json.loads(config_json)
+        # Convert json to a launch config
+        # TODO Implement launch config
+        lc = LaunchConfig()
+        ctx = LaunchContext(namespace='/', filename=filename)
+        if argv:
+            ctx = ctx.with_argv(argv)
+
+        ctx, cfg = self._load_launch_objects(ctx, lc,
+                                             [list(config_nodes)])
+        logger.debug(f'launch configuration: {cfg}')
+        return lc
+
+    def _load_launch_objects(self, ctx: LaunchContext,
+                             cfg: LaunchConfig,
+                             node_list: Sequence[Sequence[Dict]]
+                             ) -> Tuple[LaunchContext, LaunchConfig]:
+        for nodes in node_list:
+            for node in nodes:
+                if node['__TYPE__'] == 'Node':
+                    print(f'processing: {node}')
+                    # remappings = self.\
+                    #     _get_remappings_from_json(node.get('remappings'))
+                    args: Optional[Any] = node.get('args', [])
+                    if isinstance(args, list):
+                        args = ' '.join(args)
+                    print(args)
+                    remappings = node.get('remappings', [])
+                    if remappings is not None:
+                        remappings = tuple(remappings)
+                    nc = NodeConfig(
+                        name=node['name'],
+                        namespace=node['namespace'],
+                        package=node['package'],
+                        executable_path=node['executable_path'],
+                        executable_type=ExecutableType[
+                            node['executable_type']
+                        ],
+                        remappings=remappings, # noqa
+                        filename=node.get('filename'),
+                        output=node.get('output'),
+                        required=self._get_bool(node.get('required'), False),
+                        respawn=self._get_bool(node.get('respawn'), False),
+                        respawn_delay=self._get_float(
+                            node.get('respawn_delay'), 0.0),
+                        env_args=tuple(node.get('env_args', [])), # noqa
+                        cwd=node.get('cwd'),
+                        args=args,
+                        launch_prefix=node.get('launch_prefix'),
+                        typ=''
```

Comment:
Empty string is a bad smell (i.e., use of sentinel value). Could lead to unexpected, difficult to debug errors at run-time. We should either make `typ` optional or figure out what the ROS2 equivalent should be.

Reply:
Fixed by moving the interpretation of ExecuteProcess down into the container script. Now only Nodes are returned.